src/library/scala/collection/immutable/Vector.scala

```diff
@@ -417,11 +417,11 @@ private final class Vector1[+A](_data1: Arr1) extends VectorImpl[A](_data1) {
   protected[immutable] def vectorSlice(idx: Int): Array[_ <: AnyRef] = prefix1
   protected[immutable] def vectorSlicePrefixLength(idx: Int): Int = prefix1.length
 
-  override protected[this] def prependedAll0[B >: A](prefix: collection.IterableOnce[B], k: Int): Vector[B] = {
-    val data1b = prepend1IfSpace(prefix1, prefix)
-    if(data1b ne null) new Vector1(data1b)
-    else super.prependedAll0(prefix, k)
-  }
+  override protected[this] def prependedAll0[B >: A](prefix: collection.IterableOnce[B], k: Int): Vector[B] =
+    prepend1IfSpace(prefix1, prefix) match {
+      case null => super.prependedAll0(prefix, k)
+      case data1b => new Vector1(data1b)
+    }
```

Comment:
is there a reason to switch to pattern matching? for weird reasons that I'm sure someone other than me understands, it produces more bytecode.

<details>
<summary>REPL example:</summary>

```scala
scala> object Foo {
     |   private[this] def maybeAnyRef(): AnyRef =
     |     if (scala.util.Random.nextBoolean()) new AnyRef else null
     | 
     |   private[this] def identityHashCode(ref: AnyRef): String =
     |     s"${System.identityHashCode(ref)}"
     | 
     |   def bar(): String = {
     |     val ref = maybeAnyRef()
     |     if (ref == null) "null"
     |     else identityHashCode(ref)
     |   }
     | 
     |   def baz(): String =
     |     maybeAnyRef() match {
     |       case null => "null"
     |       case ref  => identityHashCode(ref)
     |     }
     | }
object Foo

scala> :javap -c -filter Foo
Compiled from "<console>"
public class Foo$ {
  public java.lang.String bar();
    Code:
       0: aload_0
       1: invokespecial #53                 // Method maybeAnyRef:()Ljava/lang/Object;
       4: astore_1
       5: aload_1
       6: ifnonnull     14
       9: ldc           #55                 // String null
      11: goto          19
      14: aload_0
      15: aload_1
      16: invokespecial #57                 // Method identityHashCode:(Ljava/lang/Object;)Ljava/lang/String;
      19: areturn

  public java.lang.String baz();
    Code:
       0: aload_0
       1: invokespecial #53                 // Method maybeAnyRef:()Ljava/lang/Object;
       4: astore_2
       5: aload_2
       6: ifnonnull     15
       9: ldc           #55                 // String null
      11: astore_1
      12: goto          27
      15: goto          18
      18: aload_0
      19: aload_2
      20: invokespecial #57                 // Method identityHashCode:(Ljava/lang/Object;)Ljava/lang/String;
      23: astore_1
      24: goto          27
      27: aload_1
      28: areturn

  public Foo$($iw);
    Code:
       0: aload_0
       1: invokespecial #27                 // Method java/lang/Object."<init>":()V
       4: return
}
```
</details>

Reply:
Thanks sjrd! I was not aware. I can confirm that with scala 2, the match adds a local even under `-opt`, but with sjrd's branch the bytecode is identical (except that the null check is inverted). Without `-opt`, match incurs `astore 4`.

The other twist is that `data1b` as a pattern variable does not get an entry in local var table.

It seems dotty currently incurs `astore 4` just around the usage of `data1b` in `new Vector1(data1b)`; maybe for that reason, it shows up in the local var table. Otherwise it's as clean as sjrd scala 2.

Reply:
Also thanks NthPortal for taking a look. There may be another discussion to have about the idiom as style, that is, "abuse of match" to replace if/else. This is not as egregious as `case _ if b`; the other use case is to replace `isInstanceOf`/`asInstanceOf` pairs, which some people prefer (for reasons that escape me, unless it is for the same concern about "bloated code"). Briefly, I see that in this case, match avoids introducing a variable that is nullable. But really I just wanted to eliminate the enclosing braces.

Reply:
> The other twist is that `data1b` as a pattern variable does not get an entry in local var table.

I think that's https://github.com/scala/scala-dev/issues/811